springernature/stylelint-config-springernature
# Stylelint plugin

# Description

Sass-lint is being used in several FE projects in Springer-Nature as the Frontend book indicates ([reference](https://github.com/springernature/sasslint-config-springernature))

The project Sass-lint used by SN is deprecated with a last stable release in 2017 ([reference](https://github.com/sasstools/sass-lint)). At the same time, the version of Sass used (node-sass) is also deprecated as 2020 ([reference](https://sass-lang.com/blog/libsass-is-deprecated/)). 

The current PR for the migration has been created and tested in the [ICE Librarian project](https://github.com/springernature/snap). 

Stylelint is not just a SCSS but a CSS linter that helps you avoid errors and enforce conventions. It ss currently being developed with the latest version (15.11.0)  released last week. It has a healthy development community as well as multiple plugins. 4,220,219 weekly downloads, and MIT License. ([reference](https://www.npmjs.com/package/stylelint))

## Plugin vs Config

Stylint allows you to create plugins that extend existing configuration and to extend them. In order to have a more versatile configuration that allows easy migration, we created a Springer Nature plugin. The advantages of this are:

- We can install the plugin with the only direct dependency of our project being stylelint. This package will look the config file automatically and will connect it with our plugin without extra configuration.
- We don't need to point to the config file inside node_modules anymore.
- In the case that is needed (although not recommended) we can extend this plugin with new rules, or disabling existing rules.

## Process for the upgrade

The following process is required to migrate from `sass-lint` to `stylelint`:

- Install stylelint in the project.
- Create a styleint configuration file according to the README.
- Remove `sass-lint` and any other related packages.
- Install the package for the build system the project is using. For example `stylelint-gulp`.
- Remove the current config and scripts in package.json related to `sass-lint`.
- Create new scripts and process for stylelint.
- Test the new scripts
- Automatic fix for the whole project
- Manual fix of the SCSS of the project to get a clean lint report.
- Upgrade of `node-sass `(depecrated) to dart-sass if needed.
- Run visual tests to check that nothing changed on the FE.
- Install and configure Stylelint on VsCode or favourite IDE.


## Next Step

- Divide the rules between `Stylelint` and `Prettier`.
- Move to the same location `Styleling`, `Eslint` and `Prettier` plugins/configs.
- Analyse and implement new SCSS configurations.
- Celebrate.
